### ISSUE TITLE
When coroutine stacks overflow kill the query, not the server

### ISSUE THREAD

**Issue by coffeemug** [rethinkdb/rethinkdb#2639]
Take a query like this:

```
r.expr({id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:{id:1}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}})
```

This one works now, but increase the depth arbitrarily (to 180 as of this writing). The server crashes:

```
error: Error in ../src/arch/runtime/thread_pool.cc at line 341:
error: Callstack overflow in a coroutine
error: Backtrace:
error: Wed Jun 25 14:48:28 2014

       1: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:101
       2: backtrace_t::backtrace_t() at backtrace.cc:202
       3: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:282
       4: format_backtrace(bool) at backtrace.cc:197
       5: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
       6: linux_thread_pool_t::sigsegv_handler(int, siginfo*, void*) at thread_pool.cc:341
       7: /lib/x86_64-linux-gnu/libpthread.so.0(+0xfcb0) [0x7f5536b57cb0] at 0x7f5536b57cb0 (/lib/x86_64-linux-gnu/libpthread.so.0)
       8: google::protobuf::internal::MessageTypeTraits<Backtrace>::Get(int, google::protobuf::internal::ExtensionSet const&, Backtrace const&) at extension_set.h:713
       9: google::protobuf::internal::MessageTypeTraits<Backtrace>::ConstType Term::GetExtension<google::protobuf::internal::MessageTypeTraits<Backtrace>, (unsigned char)11, false>(google::protobuf::internal::ExtensionIdentifier<Term, google::protobuf::internal::MessageTypeTraits<Backtrace>, (unsigned char)11, false> const&) const at ql2.pb.h:1722
       10: ql::get_backtrace(ql::protob_t<Term const> const&) at error.cc:127
       11: ql::term_t::term_t(ql::protob_t<Term const>) at term.cc:328
       12: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at op.cc:184
       13: ql::make_obj_term_t::make_obj_term_t(ql::compile_env_t*, ql::protob_t<Term const> const&) at datum_terms.cc:60
       14: counted_t<ql::make_obj_term_t> make_counted<ql::make_obj_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at counted.hpp:141
       15: ql::make_make_obj_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at datum_terms.cc:91
       16: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at term.cc:23
       17: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at op.cc:207
       18: ql::make_obj_term_t::make_obj_term_t(ql::compile_env_t*, ql::protob_t<Term const> const&) at datum_terms.cc:60
       19: counted_t<ql::make_obj_term_t> make_counted<ql::make_obj_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at counted.hpp:141
       20: ql::make_make_obj_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at datum_terms.cc:91
       21: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at term.cc:23
       22: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at op.cc:207
...
```

Since it's really hard in general to look at user input and figure out whether it will cause a stack overflow, we should detect stack overflows and recover cleanly with an error for the user.

This was originally discussed in #173.

**Comment by neumino** [comment]
So this thing can now happen if you open a change feed on a table and insert the changes in the same table. You can deeply nested document, and it eventually overflow.

It's probably not a real use case, but I was just testing writable streams on rethinkdbdash, and the idea of plugging a table on itself was too funny to not try :)

**Comment by coffeemug** [comment]
Err. We should definitely fix that.

**Comment by danielmewes** [comment]
For most practical cases, this will be addressed by the fix described in https://github.com/rethinkdb/rethinkdb/issues/4024#issuecomment-98884353 that is now in CR 2955.

**Comment by danielmewes** [comment]
Fixed in v2.0.x as of https://github.com/rethinkdb/rethinkdb/commit/195f365c4e9473de5bfeb2d3bcc8dde488c37c55 and cherry-picked into next in https://github.com/rethinkdb/rethinkdb/commit/4ad7aa042c0bafe247e88d5e5324ad4b1b4d8e90 and https://github.com/rethinkdb/rethinkdb/commit/9ca2bacb43ed4464e02db65ea66d4fc02aab7afa

We now check that we have at least 4 KB of stack space available before compiling or evaluating another term, and fail with a ReQL exception if that's not the case.
